qubesadmin/tools/qvm_shutdown.py

```diff
@@ -48,37 +48,66 @@
     help='timeout after which domains are killed when using --wait'
         ' (default: %d)')
 
+def distance_from_netvm(vm):
+    ''' Calculate distance from netvm chain root.
+
+    VM not connected to any netvm get 0, a VM connected to it get 1, etc.
+    '''
+    distance = 0
+    try:
+        while vm.netvm is not None:
+            distance += 1
+            vm = vm.netvm
+    except (KeyError, AttributeError):
+        # no netvm, or we don't have access to it
+        pass
+    return distance
 
 def main(args=None, app=None):  # pylint: disable=missing-docstring
     args = parser.parse_args(args, app=app)
 
-    for vm in args.domains:
-        try:
-            vm.shutdown(force=args.force)
-        except qubesadmin.exc.QubesVMNotStartedError:
-            pass
-
-    if not args.wait:
-        return
-
-    timeout = args.timeout
-    current_vms = list(sorted(args.domains))
-    while timeout >= 0:
-        current_vms = [vm for vm in current_vms
-            if vm.get_power_state() != 'Halted']
-        if not current_vms:
-            return 0
-        args.app.log.info('Waiting for shutdown ({}): {}'.format(
-            timeout, ', '.join([str(vm) for vm in current_vms])))
-        time.sleep(1)
-        timeout -= 1
-
-    args.app.log.info(
-        'Killing remaining qubes: {}'
-        .format(', '.join([str(vm) for vm in current_vms])))
-    for vm in current_vms:
-        vm.kill()
-
+    # care about order only when --wait is used:
+    if args.wait:
+        domains_and_distance = [(vm, distance_from_netvm(vm))
```

Comment:
It is more complex than this - if `vm.shutdown()` raise an exception, VM should go to the next round. But if timeout, it should be killed, at the end of the same round. Otherwise for example hanging `untrusted` VM would block attempt to clean shutdown `sys-net` and `sys-firewall`.